File: dynamic_programing/java/Min_Coins.java
public class CoinChangeDyn {
	
	public static void main(String args[]) {
		int d[] = { 1, 5, 8, 10, 25 }; // Coins
		int n = d.length;
		int c = 33; // Change wanted
		int t[] = new int[c + 1];
		for (int i = 1; i <= c; i++) {
			t[i] = i;
			for (int j = 1; j < n; j++) {
				if (i >= d[j]) {
					t[i] = Math.min(t[i],  1 + t[i - d[j]]);
				}
			}
		}
		
		// Outputs the number of coins needed
		System.out.println(t[c]);
	}
}

File: dp/min_coins/python3/min_coins.py
def coinChange(coins, amount):
    if amount == 0:
        return 0
    if min(coins) > amount:
        return -1
    dp = [-1 for i in range(0, amount + 1)]
    for i in coins:
        if i > len(dp) - 1:
            continue
        dp[i] = 1
        for j in range(i + 1, amount + 1):
            if dp[j - i] == -1:
                continue
            elif dp[j] == -1:
                dp[j] = dp[j - i] + 1
            else:
                dp[j] = min(dp[j], dp[j - i] + 1)
    return dp[amount]


print(coinChange([1, 5, 10, 25, 50], 199))
